src/selectors_vm/compiler.rs

```diff
@@ -23,46 +31,134 @@ pub struct CompiledAttributeExprOperand {
                 if $negation {
-                    Box::new(move |$arg| !($body))
+                    Box::new(move |$state, $arg| !($body))
                 } else {
-                    Box::new(move |$arg| $body)
+                    Box::new(move |$state, $arg| $body)
                 }
             };
 
             (@unmatchable) => {
-                compile_expr!(|_arg| false);
+                compile_expr!(|_, _| false)
             };
 
-            (@match_all) => {
-                compile_expr!(|_arg| true);
-            };
+            (@match_any) => {
+                compile_expr!(|_, _| true)
+            }
         }
-    };
+    }
 }
 
-trait CompileOr<T> {
-    fn compile_or(
-        self,
-        negation: bool,
-        compile: impl Fn(T) -> CompiledAttributeExpr,
-    ) -> CompiledAttributeExpr;
+trait Compilable {
+    fn compile(&self, encoding: &'static Encoding, exprs: &mut ExprSet, flags: &mut ProgramFlags);
 }
 
-impl<T> CompileOr<T> for Result<T, ()> {
-    #[inline]
-    fn compile_or(
-        self,
-        negation: bool,
-        compile: impl Fn(T) -> CompiledAttributeExpr,
-    ) -> CompiledAttributeExpr {
+impl Compilable for Expr<OnTagNameExpr> {
+    fn compile(
+        &self,
+        encoding: &'static Encoding,
+        ExprSet { local_name_exprs, .. }: &mut ExprSet,
+        flags: &mut ProgramFlags,
+    ) {
+        let &Self { negation, ref simple_expr } = self;
         curry_compile_expr_macro!(negation);
 
-        match self {
-            Ok(v) => compile(v),
-            Err(_) => compile_expr!(@unmatchable),
+        local_name_exprs.push(
+            match simple_expr {
+                OnTagNameExpr::ExplicitAny => compile_expr!(@match_any),
+                OnTagNameExpr::Unmatchable => compile_expr!(@unmatchable),
+                OnTagNameExpr::LocalName(local_name) => {
+                    match LocalName::from_str_without_replacements(&local_name, encoding)
+                        .map(LocalName::into_owned)
+                    {
+                        Ok(local_name) => {
+                            compile_expr!(|_, actual_local_name| *actual_local_name == local_name)
+                        }
+                        // NOTE: selector value can't be converted to the given encoding, so
+                        // it won't ever match.
+                        Err(_) => compile_expr!(@unmatchable),
+                    }
+                }
+                &OnTagNameExpr::NthChild(nth) => {
+                    compile_expr!(|state, _| state.counter.is_nth(nth))
+                }
+                &OnTagNameExpr::NthOfType(nth) => {
+                    *flags |= ProgramFlags::NTH_OF_TYPE;
+                    compile_expr!(|state, _| state.type_counter.expect("Counter for type required at this point").is_nth(nth))
+                }
+            }
+        );
+    }
+}
+
+impl Compilable for Expr<OnAttributesExpr> {
+    fn compile(
```

Comment:
I find this variant of the function much harder to read. What we want to achieve with this refactoring?

Reply:
It made more sense when there were more ordered expression lists and separate "ops" but now it doesn't seem that worth it now that everything is more or less the same as before.

Reply:
Actually, I just noticed how we can remove the "curry" and "compile" macros which made compilation look a lot more complicated like I originally wanted to do with this refactor.